### PR TITLE
fix certain broken regular expressions by quoting

### PR DESCRIPTION
**Pull request by ijsnow**
This PR addresses https://github.com/sourcegraph/sourcegraph/issues/2125. The query parser now places quotation marks around search queries (treats as a string literal) if they are a regular expression and are invalid for one of the following reasons:

- There is an unclosed `(`
- There is an unclosed `[`

Closes https://github.com/sourcegraph/sourcegraph/issues/2125.

The issue description currently says that we should also treat queries with unclosed `{` as well. However, I have determined that we do not need to do this. `foo{` is a valid regular expression and behaves as expected. [See this commit/test for proof.](https://github.com/sourcegraph/sourcegraph/pull/2999/commits/65b600ca741d94e8170b65e12a69f3bda8405ff7)

Before merging, we should revert https://github.com/sourcegraph/sourcegraph/pull/2999/commits/65b600ca741d94e8170b65e12a69f3bda8405ff7.

Test plan: A test was added to the code.

### CODE REVIEW

File: cmd/frontend/internal/pkg/search/query/types/check.go [hunk 1]
```diff
@@ -105,27 +105,35 @@ func (c *Config) checkExpr(expr *syntax.Expr) (field string, fieldType FieldType
...
 	case StringType:
 		dst.String = &valueString
 	case RegexpType:
 		p, err := regexp.Compile(valueString)
 		if err != nil {
-			return err
+			switch field {
+			case "":
+				if shouldQuoteQuery(err) {
+					return setValue(dst, `"`+valueString+`"`, StringType, field)
```
Comment (keegancsmith) [inline cmd/frontend/internal/pkg/search/query/types/check.go]: This seems like a very naive way to quote a string. Why not just use `regexp.QuoteMeta` and Compile that regex? You can then also avoid adding field as a parameter.

**Comment by ijt** [comment]
> From a product POV, I worry that this will confuse the user when they get no error message for invalid regexps like `foo([a-z+)` (where they forgot the `[`). If a user types in that regexp, 99.999% of the time they would want an error message, not to have their regexp silently treated literally. Do you agree?
> 
> Instead, we should ONLY fix up the cases where the user **almost certainly** made a mistake. For example, if a user types in `foo(`, 99.999% of the time they would want it to be interpreted as `foo\(`, and 0.001% of the time would they want an error message.
> 
> Also, how does this relate to #2991? They both address the same problem. Did you intend for both PRs to be merged, or are you looking for product or code review input to determine which one is better?

I independently started work on #2991 thinking that @ijsnow was working on search refactoring. It was a failure of communication. Only one of these two PRs should be merged.

**Comment by ijt** [comment]
One thing to consider from a UX perspective is what happens when someone enters for example `repo:ph/sourcegraph error(` but only wanted to see calls to functions and methods called `error` without any prefix. Too many results show up, so the user puts `\b` on the front, resulting in `repo:ph/sourcegraph \berror(`. With the quoting approach, this new query fails to return anything. With the approach of only escaping the paren like `error\(` it still works fine because it turns into `\berror\(`.

**Comment by sqs** [comment]
Can we close this in favor of #2991? I think that approach is better for the reasons @ijt and I described above.

**Comment by keegancsmith** [comment]
I re-reviewed #2991. I agree making the cases we autofix super limited is the best approach. #2991 now implements that so this PR can be abandoned.

**Comment by sqs** [comment]
Abandoning per comments above

**Pull request closed without merging**